### DISCUSSION TITLE
JSON Encoding and Decoding Functions

### DISCUSSION THREAD

**Discussion by springcomp** [jmespath-community/jmespath.spec#22] [JEP]
# JSON Functions (GorillaStack)

|||
|---|---
| **JEP**    | 17
| **Author** | Chris Armstrong (Gorilla Stack)
| **Status** | draft
| **Created**| 31-10-2019

## Abstract

This document proposes adding functions for manipulating JSON data, including
a `parse_json` and `to_json` function.

## Motivation

There is separate requirements to both parse JSON strings and to serialise
full or partial object structures to JSON.

## Specification

The following functions will be added to the specification:

### parse_json

```
any parse_json(string $json_data)
```

Parse string `$json_data` containing JSON data and produce an object structure that can be
further manipulated by JMESPath.

If the string contains invalid JSON, an error should be generated.

### to_json

```
string to_json(any $value)
```

Serialise `$value` to a JSON string. No guarantees are made about formatting - this is
implementation dependent.

## Impact

Although this introduces no changes to existing syntax or functions, it may be challenging
to support correctly in implementations other than JavaScript if the underlying language
does not contain a spec-compliant JSON parser or producing arbitrary object structures for
use with JMESPath is difficult.

**Comment by springcomp** [comment]
@chris-armstrong I took the liberty to include your proposal here for the next version of JMESPath.

Would you be so kind to elaborate one or two use cases that lead you to create those functions and that you could not do using JMESPath as currently specified ?

**Reply by chris-armstrong** [reply]
The use cases it was devised for was:

 to manipulate JSON encoded strings (inside existing JSON Object payloads)

to allow a manipulated JMESPath object to be serialiser into a JSON string for inclusion in a payload where it was expected

(apologies for not being more specific, it’s been three years since this proposal was submitted and I no longer use JMESPath, although @gorillastack most likely still does)

**Reply by springcomp** [reply]
Thanks a lot for your feedback.

OK, so that would be if you have double-encoded JSON. like so:

- decode: `` search( parse_json(foo).bar, { "foo": "{\"bar\": \"baz\"}" } ) -> "baz" ``
- encode: `` search( to_json(foo), { "foo": { "bar": "baz" } } ) -> "{\"bar\": \"baz\"}" ``

Am I correct?

**Reply by chris-armstrong** [reply]
Yes this is its intended usage
